src/main/java/org/zalando/baigan/proxy/handler/ContextAwareConfigurationMethodInvocationHandler.java

```diff
@@ -107,13 +107,16 @@ private Object getConfig(final String key, final List<ContextProvider> contextPr
         }
 
         if (!CollectionUtils.isEmpty(contextProviders)) {
-            if (contextProviders.size() > 1) {
-                LOG.warn("The key [{}] has more than one context provider, and therefore only the first context will be used", key);
-            }
-            final ContextProvider contextProvider = contextProviders.get(0);
-            contextProvider
-                    .getProvidedContexts()
-                    .forEach(param -> context.put(param, contextProvider.getContextParam(param)));
+            contextProviders.forEach(contextProvider -> {
+                contextProvider
+                        .getProvidedContexts()
+                        .forEach(contextParam -> {
+                            if(context.containsKey(contextParam)){
+                                throw new RuntimeException("Cannot have more than one context provider for the same context key "+contextParam);
+                            }
+                            context.put(contextParam, contextProvider.getContextParam(contextParam));
+                        });
+            });
```

Comment:
Just noting: if we used the reflection approach suggested by @lukasniemeier-zalando , this check could be done on context creation instead of at query time, making this more efficient and fail earlier.